Improvement for the S3 support - IAM Role and support for legacy folder structure

vins01-4science opened the pull request:
## References
* Fixes #8363

## Description
This PR introduces some improvements and a fixes for the S3Bucket AssetStore:
- Fixes Amazon S3Bucket AssetStore issue;
- Enables authentication using IAM roles;
- Enables saving buckets using different subpaths .

## Instructions for Reviewers
List of changes in this PR:
* Improved `S3BitStoreService`;
* Added:
    -  UnitTest for `S3BitStoreService` using `Mockito`;
    - `FunctionalUtils` class that contains some functional utilities;
    - `storage.cfg` configuration file.
* Modified configuration Files (`dspace.cfg`, `local.cfg.EXAMPLE`, `bitstore.cfg`);

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

abollini committed:
DSC-403 Porting to DSpace-CRIS 7 of the IAM support for S3

Micheleboychuk committed:
added support to use relative path in S3BitStore

Micheleboychuk committed:
refactored s3bitStore

Commit:
[DSC-403] Fixed S3 use relative path property name

vins01-4science committed:
[DURACOM-69] Added UnitTest
feat: Added FunctionalUtils that uses Functional Interfaces and declares some utilities
fix & imp: S3BitStoreService fixed some issues and refactored code
test: Added S3BitStoreServiceTest for missing UnitTest

vins01-4science committed:
Merged main into main-duracom-69

vins01-4science committed:
[DURACOM-69] Refacored common methods

ref: Introduced abstract common class with utility methods

vins01-4science committed:
Merge remote-tracking branch 'bit-bucket/main-duracom-69' into main-duracom-69

vins01-4science committed:
Merge remote-tracking branch 'origin/main' into main-duracom-69

tdonohue commented on dspace-api/src/main/java/org/dspace/util/FunctionalUtils.java:
```diff
@@ -0,0 +1,57 @@
+ * @author Vincenzo Mecca (vins01-4science - vincenzo.mecca at 4science.com)
+ *
+ */
+public class FunctionalUtils {
```
Please add JavaDocs to describe.  For instance, how are `FunctionalUtils` different from other utility classes under `org.dspace.utils.*` or even `org.dspace.core.Utils`

vins01-4science replied:
As described with the `JavaDocs`:
> These methods are linked to the functional paradigm and use {@code Functional} interfaces of java 8+, all the main
> interfaces are in the package {@link java.util.function}.

tdonohue commented on dspace/config/modules/storage.cfg:
```diff
@@ -0,0 +1,31 @@
+#---------------------------------------------------------------#
+#-----------------STORAGE CONFIGURATIONS------------------------#
+#---------------------------------------------------------------#
+# Configuration properties used by the bitstore.xml config file #
```
It's confusing to me that this file includes everything *except* `assetstore.dir` (which is the main configuration needed for `localStore`).

Should we consider renaming this file to be `assetstore.cfg` (as all configs start with `assetstore.*` and also MOVING `assetstore.dir` to this file?  We could simply leave a comment in `dspace.cfg` where `assetstore.dir` exists that says: 

```
# Assetstore configurations have moved to config/modules/assetstore.cfg
# and config/spring/api/bitstore.xml.
```

tdonohue commented on dspace/config/modules/storage.cfg:
```diff
@@ -0,0 +1,31 @@
+#---------------------------------------------------------------#
+
+# Use the localStore or the s3Store implementation
+assetstore.storename.0 = localStore
```
I'm not a fan of ending this with `.0` because it implies you can simply add an `assetstore.storename.1` and expect it to work.  It won't work though as if you want multiple they must be specified in `bitstore.xml`.

I'd recommend we change this to better describe the setting to be something like:
```
# Use 'localStore' or 's3Store' as the primary assetstore. 
# These named assetstores are defined in config/spring/api/bitstore.xml
# If you want to add additional assetstores, they must be added to that bitstore.xml
assetstore.storename.primary = localStore
```

tdonohue commented on dspace/config/modules/storage.cfg:
```diff
@@ -0,0 +1,31 @@
+assetstore.storename.0 = localStore
+
+# For using a relative path (xx/xx/xx/xxx...) set to true, default it false
+assetstore.s3.useRelativePath = false
```
I'm not sure I understand this configuration. Could we describe it more clearly?  Where does this "relativePath" come from?  Can we give an example in the comments?

tdonohue commented on dspace/config/modules/storage.cfg:
```diff
@@ -0,0 +1,31 @@
+# Use the localStore or the s3Store implementation
+assetstore.storename.0 = localStore
+
+# For using a relative path (xx/xx/xx/xxx...) set to true, default it false
```
Starting with this config, we should have a header here that says something like:
```
#---------------------------------------------------------------#
#------------ Amazon S3 Specific Configurations -----------#
#---------------------------------------------------------------#
# The below configurations are only used if the primary storename
# is set to 's3Store' or the 's3Store' is configured as a secondary store
# in your bitstore.xml
```

tdonohue commented on dspace/config/modules/storage.cfg:
```diff
@@ -0,0 +1,31 @@
+
+# S3 bucket name to store assets in, default would generate a bucket
+# based on the dspace host name
+assetstore.s3.bucketName =
```
Reword to say something like:
```
# S3 bucket name to store assets in. If unspecified, by default DSpace will
# create a bucket based on the hostname of `dspace.ui.url` setting.
```

tdonohue commented on dspace/config/modules/storage.cfg:
```diff
@@ -0,0 +1,31 @@
+assetstore.s3.subfolder =
+
+# please do not use these in production but rely on the aws credentials
+# discovery mechanism to configure them (ENV VAR, EC2 Iam role, etc.)
```
Can we provide more information about the recommended approach to setting these? It seems very odd to have configurations which we don't recommend using. But, if we are doing so, we need to provide them with links on how to set these via environment variables, etc.  Or give examples of what environment variables to user to set these.

tdonohue commented on dspace/config/modules/storage.cfg:
```diff
@@ -0,0 +1,31 @@
+assetstore.s3.awsSecretKey =
+# to force the use of a specific region when credentials are provided
+# in this configuratin file. If credentials are left empty this prop
+# is ignored
```
I don't think this description is accurate.  From the code, it looks like `awsRegionName` is set to the AWS `DEFAULT_REGION` (which I think is `us-east-1`?) if it's not specified.

vins01-4science replied:
Yes, the `awsRegionName` is configured with the `DEFAULT_REGION` if the property is left blank.
But the comment
> to force the use of a specific region when credentials are provided
> in this configuratin file. **If credentials are left empty this prop
>  is ignored**
assures that if credentials are empty the region is ignored, because we will use the default credential manager of AWS

tdonohue replied:
@vins01-4science : I'd recommend describing that better then.  I do not think the current text is clear.  Maybe say something like:

`If the credentials are left empty, then this setting is ignored and the default AWS region will be used.`

tdonohue commented on dspace/config/local.cfg.EXAMPLE:
```diff
@@ -247,3 +247,30 @@ db.schema = public
+
+#---------------------------------------------------------------#
+#-----------------STORAGE CONFIGURATIONS------------------------#
+#---------------------------------------------------------------#
```
I would recommend removing all of these from `local.cfg.EXAMPLE`.  We are trying to keep this example file minimal...notice it doesn't list every configuration.

Instead, if you want to mention S3 in this file, I'd recommend adding a comment above the `assetstore.dir` setting in `local.cfg.EXAMPLE` to say something like:
```
# Additional storage options (e.g. Amazon S3) are available in `assetstore.cfg`
```

tdonohue reviewed (changes requested):
@vins01-4science : Thanks for this update/improvement to S3.  Overall the code looks good, but I have some recommendations below on how to make this easier to understand & use  (mainly some cleanup of the configurations and moving them all to an `assetstore.cfg`).

I'm not able to fully test this myself right now as I don't have an S3 account ready to go.  But, I did verify that the refactors to the localStore work fine.  I'm hoping @benbosman will find a chance to test the S3 side of things in the coming weeks.

hardyoyo commented:
I love the idea of this PR, please let me know if there's anything I can do to help

vins01-4science committed:
[#8356][DURACOM-69] Change Requests

vins01-4science committed:
[DSpace#8356][DURACOM-69] Change Requests part. 2

vins01-4science commented:
Hello @tdonohue,
now all change requests should be resolved.
Thank you.

tdonohue commented on dspace-api/src/main/java/org/dspace/storage/bitstore/BaseBitStoreService.java:
```diff
@@ -0,0 +1,224 @@
+     * @param internalId The internal_id
+     * @return The path based on the id without leading or trailing separators
+     */
+    protected String getIntermediatePath(String internalId) {
```
While not required, it might be nice to add some Unit tests for this refactored method to ensure it's still operating the same as before. There is a lot of code refactoring in this method, and while it looks right, it's hard to tell at a glance.  It'd be better to have tests which verify that for given `internalId` value(s) we end up with the intermediate path we expect.

vins01-4science replied:
New tests added in `S3BitStoreServiceTest`.

tdonohue commented on dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:
```diff
@@ -54,10 +62,13 @@ public class S3BitStoreService implements BitStoreService {
      */
     private static final String CSA = "MD5";
+    protected static final int digitsPerLevel = 2;
+    protected static final int directoryLevels = 3;
```
We should add the same warnings here as we have in `BaseBitStoreService`:
```
// These settings control the way an identifier is hashed into
    // directory and file names
    //
    // With digitsPerLevel 2 and directoryLevels 3, an identifier
    // like 12345678901234567890 turns into the relative name
    // /12/34/56/12345678901234567890.
    //
    // You should not change these settings if you have data in the
    // asset store, as the BitstreamStorageManager will be unable
    // to find your existing data.
```

tdonohue commented on dspace/config/local.cfg.EXAMPLE:
```diff
@@ -247,3 +247,8 @@ db.schema = public
+#---------------------------------------------------------------#
+#-----------------STORAGE CONFIGURATIONS------------------------#
+#---------------------------------------------------------------#
+# Additional storage options (e.g. Amazon S3) are available in `assetstore.cfg`
```
I'd recommend removing all these notes at the end of local.cfg.EXAMPLE (i.e. remove lines 250-254).  

Instead, copy your changes to dspace.cfg over to lines 50-52 above.  That way lines 50-52 in `local.cfg.EXAMPLE` are replaced with
```
# Assetstore configurations have moved to config/modules/assetstore.cfg
# and config/spring/api/bitstore.xml.
# Additional storage options (e.g. Amazon S3) are available in `assetstore.cfg`
# assetstore.dir = ${dspace.dir}/assetstore
```

tdonohue commented on dspace/config/modules/assetstore.cfg:
```diff
@@ -0,0 +1,49 @@
+# please don't use root credentials in production but rely on the aws credentials default
+# discovery mechanism to configure them (ENV VAR, EC2 Iam role, etc.)
+# The preferred approach for security reason is to use the IAM user credentials, but isn't always
+possible.
```
This line is not commented out. Maybe move this word to the previous line?

tdonohue reviewed (changes requested):
@vins01-4science : This is looking much better now.  However, I still have a few minor suggestions (one of which was overlooked previously). 

I still don't have a way to easily test this, but the code is looking good.  I'm hoping @benbosman will find some time to run some tests on his end.

vins01-4science committed:
[DSpace#8356][DURACOM-69] Change Requests part. 3

vins01-4science commented:
Hello @tdonohue,
now all change requests should have been resolved.
Thank you.

tdonohue commented on dspace/config/local.cfg.EXAMPLE:
```diff
@@ -247,3 +247,8 @@ db.schema = public
+# Assetstore configurations have moved to config/modules/assetstore.cfg
+# and config/spring/api/bitstore.xml.
+# Additional storage options (e.g. Amazon S3) are available in `assetstore.cfg`
+# assetstore.dir = ${dspace.dir}/assetstore
```
@vins01-4science : You seem to have misunderstood my priori request here.  We now have two `assetstore.dir` commented out sections in `local.cfg.EXAMPLE`.  There's this one, and the one on lines 50-52 above.

I'd recommend moving these lines up to *replace* lines 50-52 in this file. It's confusing to have `assetstore.dir` instructions in two places in the same file.

vins01-4science replied:
I'm sorry for the mistake, now should be correct.
Thank you.

vins01-4science committed:
[DSpace#8356][DURACOM-69] Change Requests part. 3.1

tdonohue reviewed (approved):
👍 Thanks @vins01-4science .  All my prior feedback has been addressed now & the code looks good.

That said, I *don't* have a great place to give this a test run myself.  So, I'm going to refrain from merging this until either @benbosman or maybe @hardyoyo has a chance to give it a test run.

benbosman reviewed (changes requested):
Thanks, this looks good, only one inline comment that looks risky

vins01-4science commented on dspace/config/spring/api/bitstore.xml:
```diff
@@ -7,7 +7,7 @@
         <property name="stores">
             <map>
-                <entry key="0" value-ref="localStore"/>
+                <entry key="0" value-ref="${assetstore.storename.primary}"/>
```
Ok, so Do you suggest to add a new property inside the `dspace.cfg`?
Which should be its naming? @benbosman @tdonohue

benbosman replied:
I would rather add `${assetstore.storename.primary}` to `dspace.cfg`, set it to 0 by default, and add some comments that it should be set to 1 if they want to switch to S3

That won't break the repo if there are bitstreams already

tdonohue replied:
@benbosman : I'm not sure I understand your recommendation here.  I don't think it's easy to swap between `0` and `1` stores like you are suggesting.

Currently the `1` store is commented out by default: https://github.com/DSpace/DSpace/blob/b9eb369275cffbeb9d51d40fe6104450c7df2a36/dspace/config/spring/api/bitstore.xml#L11

Unless I'm mistaken, I don't think we can enable (uncomment) that `1` store by default without causing errors....as soon as you uncomment that `<entry>` tag, I think the `s3Store` will attempt to load & that'll fail unless you actually have AWS keys configured.

I think the overall goal of this PR was to simply allow the value of the main `0` store to be swapped easily from `localStore` to `s3Store` based on whether you wanted to primarily use local storage or S3.  If you wanted to use *both* then you are correct you'd still have to modify the `bitstore.xml` file manually to add a separate `1` store with the settings you desire.

tdonohue replied:
@benbosman and @vins01-4science :  Although, I *do* now see Ben's point that this scenario is problematic..

1. If you add Bitstreams to DSpace first, they will be stored in `localStore` by default.
2. If you then change your `assetstore.storename.primary = s3Store`, then all existing Bitstreams in `localStore` will no longer be found.  This is because the primary store is switched to use S3...and all Bitstreams will be assumed to be in S3.

So, there is a bug here because the if you've already added some content, then that existing content could be lost unless you change `bistore.xml` to have *two stores* instead of one.

vins01-4science replied:
Hello @benbosman and @tdonohue , thank  you for the comments.
So what I should do here?

tdonohue replied:
@vins01-4science : I'm a little stuck myself on the best direction forward.  I see @benbosman 's point now...and his solution would work _if we can make sure these beans are lazy loaded_ (or maybe`@ConditionalOnProperty`).

Ben's correct that part of the issue here is that we are NOT changing this tag `<property name="incoming" value="0"/>`.

However, we also shouldn't load the `S3BitStoreService` bean by default, as many sites don't need that bean (and it'll throw a ton warnings if you don't have the `aws*` properties specified).

My only idea is to look at what @benbosman suggests, but also make sure the S3 bean isn't loading by default.
1. First, ensure there's a config which lets you change the `incoming` property from `0` (localstore) to `1` (s3store)
2. Likely this means uncommenting this tag by default: `<entry key="1" value-ref="s3Store"/>`
3. However, make sure the `s3Store` bean does NOT load by default. this might require some testing to figure out the best approach.  Either finding a way to lazy load that bean, or maybe add a `@ConditionalOnProperty` annotation which uses a new configuration `assetstore.s3.enabled = true/false`

Overall, though, I see @benbosman 's point that this current implementation could cause major issues if you switch to S3 _after adding content into DSpace_.  This current implementation only works well in the scenario where you have a fresh/brand new DSpace installation which needs to be configured to only use S3.

vins01-4science replied:
@tdonohue: thank you for the clear review, Do we still need some feedback from @benbosman or I can proceed with the changes you requested earlier:

> 1. First, ensure there's a config which lets you change the incoming property from 0 (localstore) to 1 (s3store)
>2. Likely this means uncommenting this tag by default: <entry key="1" value-ref="s3Store"/>
> 3. However, we need make sure the s3Store bean does NOT load by default. this might require some testing to figure out the best approach. Either finding a way to lazy load that bean, or maybe add a @ConditionalOnProperty annotation which uses a new configuration assetstore.s3.enabled = true/false

Am I missing something else?

tdonohue replied:
@vins01-4science : I see that @benbosman already gave a "thumbs up" response to my comment just above where I suggested those 3 changes.  So, I think you can proceed with making those three changes to help move this PR forward.  

One minor additional suggestion, is that I think we may want to rename the config, as `assetstore.storename.primary` implies it should be a *name* of a store.  Now we are talking about switching this to the store *number*.  So, we may want to rename it to be something like `assetstore.index.primary` and documenting that setting it to `0` means it will use local storage, while setting it to `1` means S3 (and note that additional indexes may be added to bitstore.xml as needed).

Thanks!

vins01-4science replied:
@tdonohue : Resolved and introduced a secondary store as backup while trying temp config with another store as the primary one. 
Also introduced new status for stores `disabled` and `initialized` just to lazy initialize them (if disabled) at the first use.
Let me know if there is something else.
Thank you!

vins01-4science committed:
[#8356][DURACOM-69] Improved Configuration for S3Store

feat:
	- New Properties to configure primary store;
	- New status for stores that indicates their initialization;
	- New enabled status for stores for disabling them;
	- New secondary index, used as backup for temporary configs.

tdonohue commented on dspace/config/modules/assetstore.cfg:
```diff
@@ -0,0 +1,68 @@
+# Possible values are:
+#     - 0: to use the `localStore`;
+#     - 1: to use the `s3Store`.
+assetstore.index.secondary = 0
```
@vins01-4science : I feel like this is "scope creep", as we have not previously discussed the need for a "secondary" store location. I'm not convinced that it is necessary for most sites & I worry it'd just add more confusion to our storage configuration.

I'm also worried this new feature will cause unexpected storage issues if content is stored in the wrong storage location without your realizing it.

Here's an extreme example which is now problematic:
* Suppose you want to use `s3Store` because you don't have much local storage (100GB), but you have a `localStore` backup just in case.
* Suppose Amazon S3 is temporarily down while you are importing a large amount of content (500GB)
* Based on this implementation, you won't receive an error letting you know this large upload cannot be done **until you fill up all your local storage**.  Essentially, the secondary store will be used automatically for imported content, even if you don't want any of your content stored in that secondary store.  And there's no way to then move already imported content easily from your `localStore` to `s3Store` to free up your local storage.

I'm not in favor of a secondary store location being added at this time.  I think it'd require more discussion.   Could we just remove this secondary store option?  It's not something we asked for and it's not in the original goals of this PR....so it feels like it belongs in a separate PR.

vins01-4science replied:
@tdonohue : Sure, I will remove it, but if the `s3StoreService` isn't configured correctly, when someone will switch from `local` to `s3` some bitstream could be lost.

tdonohue replied:
@vins01-4science : If `s3StoreService` is misconfigured, it'll throw an error. I think that's appropriate here, because it blocks all new bitstreams from being added until the storage settings are fixed.

Personally, I feel most people will want to know if their storage is misconfigured (by seeing the thrown error) instead of DSpace silently hiding the misconfiguration & storing the bitstreams in the "secondary" location.

I could be wrong about my view here, but as I said, I think this needs more discussion. I'm not comfortable with adding this feature into 7.4 at the last moment as we haven't talked about it at all previously.

vins01-4science replied:
@tdonohue : Yes, sure I will remove it.
Thank you.

vins01-4science replied:
@tdonohue : removed that configuration and behavior.
Thank you.

mwoodiupui replied:
I have some ideas about policy-directed bitstore selection (for replication, nearline storage of huge objects, etc.) and a "secondary store" policy could be part of it, if we ever do that.

vins01-4science committed:
[#8356][DURACOM-69] Removed wrong secondary store option

benbosman reviewed (approved):
Thanks, this looks good now 👍

tdonohue reviewed (approved):
👍 Thanks @vins01-4science !  This looks good to me now.  I've also verified that all the LGTM issues noted above are false positives:
* one is warning about usage of MD5, but we are only using it for checksums.
* The second is warning about a Boxed variable, but the code it points to is not added by this PR. 

All in all, I think this is good to merge.  We'll need to update the documentation about how to enable/configure Amazon S3 storage.  I suspect we may want a new page specific to S3 as a subpage of https://wiki.lyrasis.org/display/DSDOC7x/Storage+Layer   If you find time, you could help me get this started. Otherwise, I'll see if I can help with this next week.

vins01-4science commented:
Thank you, @tdonohue.
> All in all, I think this is good to merge. We'll need to update the documentation about how to enable/configure Amazon S3 storage. I suspect we may want a new page specific to S3 as a subpage of https://wiki.lyrasis.org/display/DSDOC7x/Storage+Layer If you find time, you could help me get this started. Otherwise, I'll see if I can help with this next week.
I hope to find some time until next week to start that new specific S3 storage page :crossed_fingers: !

hardyoyo commented:
I'm happy to help with the docs, as soon as I, erm, figure out how to configure this correctly. :-)

hardyoyo commented:
I'm still unable to figure out how to configure this correctly, to only use an IAM Role. It looks suspiciously like I *must* have a valid user access key / secret key, which is *not* how IAM Roles work. I have confirmed that the EC2 instance I'm running this on can make aws s3 commands against my bucket, without entering any credentials.

It could just be that this PR doesn't support using IAM Roles, which is fine, but we'll need to explicitly say so in the docs.

tdonohue commented:
@vins01-4science or @abollini : Could you help us document this PR?  It appears we don't have a clear view of how to setup IAM Roles to work with DSpace.  See ⬆️

hardyoyo commented:
I've updated the docs here:
https://wiki.lyrasis.org/display/DSDOC7x/Storage+Layer#StorageLayer-ConfiguringtheBitstreamStore
